### PR TITLE
Fixing a bug causing model crash by avoiding negative channel storage

### PR DESCRIPTION
**Pull request by liho745**
MOSART can be forced to crash if negative channel storage values are produced mainly due to round errors, 
when the channel storage is completely depleted in a single time step. 
The crashing was reported in several cases of E3SM fully coupled runs, 
where the kinematic waving routing method was invoked by default. 
Jon Wolfe was able to reproduce the crashing error in 
a low-res run and print out detailed information to identify the cause.
 In this fix, a numerical treatment has been implemented 
in the kinematic wave routing method, subroutine Routing_KW(), 
to ensure that any channel storage does not get completely 
depleted in a single time step.
 This treatment is physically reasonable as long as MOSART runs 
at a sub-daily time step. 
Jon Wolfe has tested this treatment in his low-res coupled run, 
and it helped avoid crashing.  

[non-BFB]  for MOSART

### CODE REVIEW

File: components/mosart/src/riverroute/MOSART_physics_mod.F90
Comment (donghuix) [inline components/mosart/src/riverroute/MOSART_physics_mod.F90:881]: I think this will result in non BFB in the MOSART simulation.

**Review by donghuix — COMMENTED** [review]
I think this modification is not BFB.

File: components/mosart/src/riverroute/MOSART_physics_mod.F90
Comment (liho745) [reply components/mosart/src/riverroute/MOSART_physics_mod.F90:881]: You are right. This is non-BFB. It won't affect the overall model simulation (hence science), but it does change the results slightly for the rare cases when a whole channel storage is completely depleted. I added the "non-BFB" label.

Comment (rljacob) [reply components/mosart/src/riverroute/MOSART_physics_mod.F90:881]: Since its potentially non-BFB, this fix would have to go in with 3.1

**Review by bpbond — APPROVED** [review]

**Comment by peterdschwartz** [comment]
Since the new maint-3.0 has been created, I can start merging this PR this week.  I'll submit tests to see how the developer test suite is impacted.

**Comment by rljacob** [comment]
This PR is third in line of the current old answer changing PRs.  So probably next week.

Does this include the changes that were in PR #6313 ?

**Comment by peterdschwartz** [comment]
Yes, the abort is still there though the PR makes further changes to the cutoff.  Here is the full if-else block
```fortran
! check for negative channel storage
    if (negchan < -1.e-10 .and. negchan >= -1.e-8) then
       write(iulog,*) 'Warning: Small negative channel storage found! ',negchan
    elseif(negchan < -1.e-8) then
       write(iulog,*) 'Error: Negative channel storage found! ',negchan
       call shr_sys_abort('mosart: negative channel storage')
    endif
```

**Comment by jonbob** [comment]
We did capture one of the tests that ran into this abort and failed due to the earlier implementation. I can verify that this change did allow that run to successfully complete.

**Comment by peterdschwartz** [comment]
Testing on Chrysalis with this PR had `ERS.r05_r05.IELM.chrysalis_intel.elm-lnd_rof_2way` FAIL

```
118:  Error: Negative channel storage found!   -9420.09754789568
118:  ERROR: mosart: negative channel storage
118: Image              PC                Routine            Line        Source
118: libpnetcdf.so.3.0  0000155554F120CA  tracebackqq_          Unknown  Unknown
118: e3sm.exe           0000000001D90CC7  shr_abort_mod_mp_         114  shr_abort_mod.F90
118: e3sm.exe           0000000001CAE0A2  mosart_physics_mo         688  MOSART_physics_mod.F90
118: e3sm.exe           0000000001B90641  rtmmod_mp_rtmrun_        2603  RtmMod.F90
118: e3sm.exe           0000000001B05EBF  rof_comp_mct_mp_r         472  rof_comp_mct.F90
118: e3sm.exe           000000000044E85B  component_mod_mp_         757  component_mod.F90
118: e3sm.exe           0000000000424BBD  cime_comp_mod_mp_        2981  cime_comp_mod.F90
118: e3sm.exe           000000000044E47A  MAIN__                    153  cime_driver.F90
118: e3sm.exe           000000000041AA22  Unknown               Unknown  Unknown
118: libc-2.28.so       000015554882FD85  __libc_start_main     Unknown  Unknown
118: e3sm.exe           000000000041A92E  Unknown               Unknown  Unknown
```

**Comment by peterdschwartz** [comment]
Expected DIFFs from `e3sm_developer`:
```
ERS.r05_r05.IELM.chrysalis_intel.elm-V2_ELM_MOSART_features (Overall: DIFF) details:
  SMS.MOS_USRDAT.RMOSGPCC.chrysalis_intel.mosart-unstructure (Overall: DIFF) details:
```

**Comment by liho745** [comment]
> Testing on Chrysalis with this PR had `ERS.r05_r05.IELM.chrysalis_intel.elm-lnd_rof_2way` FAIL
> 
> ```
> 118:  Error: Negative channel storage found!   -9420.09754789568
> 118:  ERROR: mosart: negative channel storage
> 118: Image              PC                Routine            Line        Source
> 118: libpnetcdf.so.3.0  0000155554F120CA  tracebackqq_          Unknown  Unknown
> 118: e3sm.exe           0000000001D90CC7  shr_abort_mod_mp_         114  shr_abort_mod.F90
> 118: e3sm.exe           0000000001CAE0A2  mosart_physics_mo         688  MOSART_physics_mod.F90
> 118: e3sm.exe           0000000001B90641  rtmmod_mp_rtmrun_        2603  RtmMod.F90
> 118: e3sm.exe           0000000001B05EBF  rof_comp_mct_mp_r         472  rof_comp_mct.F90
> 118: e3sm.exe           000000000044E85B  component_mod_mp_         757  component_mod.F90
> 118: e3sm.exe           0000000000424BBD  cime_comp_mod_mp_        2981  cime_comp_mod.F90
> 118: e3sm.exe           000000000044E47A  MAIN__                    153  cime_driver.F90
> 118: e3sm.exe           000000000041AA22  Unknown               Unknown  Unknown
> 118: libc-2.28.so       000015554882FD85  __libc_start_main     Unknown  Unknown
> 118: e3sm.exe           000000000041A92E  Unknown               Unknown  Unknown
> ```

Thanks for the update. A negative storage at such a magnitude appears scientifically wrong, and hence the model should stop running. I will look into the possible reasons.

**Comment by rljacob** [comment]
@liho745 any progress on this?

**Comment by liho745** [comment]
> @liho745 any progress on this?

I ran the E3SM land developer tests on compy, and also got  ERS.r05_r05.IELM.chrysalis_intel.elm-lnd_rof_2way FAIL. 

In the corresponding baseline tests, this ERS.r05_r05.IELM.chrysalis_intel.elm-lnd_rof_2way test passed. However, when looking into the E3SM.log file, I found large negative channel storage values displayed as warnings only. 

Therefore, it is likely that the lnd_rof_2way module has caused large negative channel storage values in the first place. I will need to collaborate with the others to verify my guess further and discuss the potential options to fix the issue.

**Comment by donghuix** [comment]
I found the negative channel storage in the ```lnd_rof_2way``` test is because of the shift of time step during the coupling. I changed the coupling period of mosart to coupler to be 1800s, which fix the error.

**Comment by peterdschwartz** [comment]
Thanks I'll test it out today merged to next for the mosart tests.  Quick question, outside of this test are the elm and mosart timesteps properly synced by compset or use case?

**Comment by rljacob** [comment]
Since its Friday, wait until Monday to merge this to next.

**Comment by peterdschwartz** [comment]
the `lnd_rof_2way` test reported back a DIFF after merging it to next, so this should be ready to go monday.

**Comment by rljacob** [comment]
Go ahead and merge this to next

**Comment by peterdschwartz** [comment]
Thanks for the reminder -- been a long day.

merged to next

**Comment by wlin7** [comment]
Do we have some longer term simulations to document the impact of this fix?

**Comment by liho745** [comment]
> Do we have some longer term simulations to document the impact of this fix?

Not that I am aware of. This fix is non-BFB, but I don't expect it to change any science significantly.

**Comment by rljacob** [comment]
I didn't like the description of this PR which wrapped the bug report and the fix description in to one.  They've been separated out.

**Pull request merged**
